### PR TITLE
coala: Option to print only selected bears

### PR DESCRIPTION
**Pull request by shreyans800755**
It adds option to print only selected bears using `--show-bears -bears`.

Closes https://github.com/coala/coala/issues/4651

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by shreyans800755** [comment]
I'll add the tests soon.

File: coalib/coala.py [hunk 1]
```diff
@@ -55,8 +56,22 @@ def main(debug=False):
...
+                sections, _ = load_configuration(arg_list=None)
+                local_bears = {}
+                global_bears = {}
+                for (key, value) in sections:
+                    bear_dirs = value.bear_dirs()
+                    local_bears[key], global_bears[key] = collect_bears(
+                        bear_dirs,
+                        args.bears,
+                        [BEAR_KIND.LOCAL, BEAR_KIND.GLOBAL],
+                        warn_if_unused_glob=False)
```
Comment (mrtes) [inline coalib/coala.py]: Take a look at [collect_all_bears_from_sections](https://github.com/coala/coala/blob/master/coalib/collecting/Collectors.py#L292-L311)

Basically you are just duplicating this code.
I think it would be more beneficial to refactor that function (or some other, related function) to take a bear-glob parameter. This would also help to keep code related to bear collecting in the Collectors.py module.

**Comment by mrtes** [comment]
Two things:

1.  Maybe rename `collect_all_bears_from_sections` to `collect_bears_from_sections` or something. That would clarify the change.

2. You could potentially move the remaining code in the `args.bears` branch to a new function in `ConfigurationGathering.py`.  Or maybe refactor the `get_all_bears` function, to take the glob ;)

I dont know. What do you think about that?

**Comment by shreyans800755** [comment]
@mrtes Point 2 sounds good.

**Comment by mrtes** [comment]
Alright :)
Looks good to me
ack a91bb02

**Comment by Makman2** [comment]
Coala is always written with a lowercase c ;D

**Comment by Makman2** [comment]
Also some verb in the commit message would be good, like "Add option to ..."
You could in fact leave the tag, as every commit is affecting coala itself ;)

**Comment by Makman2** [comment]
unack a91bb02

File: coalib/coala.py [hunk 2]
```diff
@@ -54,9 +54,11 @@ def main(debug=False):
         configure_logging(not args.no_color)
 
         if args.show_bears:
-            from coalib.settings.ConfigurationGathering import (
-                get_all_bears)
-            filtered_bears = get_all_bears()
+            from coalib.settings.ConfigurationGathering import get_all_bears
```
Comment (Makman2) [inline coalib/coala.py]: unnecessary change. If it fits into a single line, please do a separate commit for it :+1:
(which is defnitely reasonable, having this in a single line reads way better)

Comment (Makman2) [reply coalib/coala.py]: (I mean just the import line, not the `filtered_bears` thing :D)

File: coalib/coala.py [hunk 3]
```diff
@@ -54,9 +54,11 @@ def main(debug=False):
...
+            if args.bears:
+                filtered_bears = get_all_bears(bear_globs=args.bears)
+            else:
+                filtered_bears = get_all_bears()
```
Comment (Makman2) [inline coalib/coala.py]: -->
```
filtered_bears = (get_all_bears(bear_globs=args.bears)
                  if args.bears else
                  get_all_bears())
```

Comment (Makman2) [reply coalib/coala.py]: or even
```
kwargs = {}
if args.bears:
    kwargs['bear_globs'] = args.bears
filtered_bears = get_all_bears(**kwargs)
```

File: coalib/collecting/Collectors.py [hunk 4]
```diff
@@ -289,12 +289,15 @@ def get_all_bears_names():
     return [bear.name for bear in get_all_bears()]
 
 
-def collect_all_bears_from_sections(sections, log_printer=None):
+def collect_all_bears_from_sections(sections,
+                                    log_printer=None,
+                                    bear_globs=['**']):
```
Comment (Makman2) [inline coalib/collecting/Collectors.py]: if possible, try to use a tuple so the default value is immutable :+1:

File: coalib/settings/ConfigurationGathering.py [hunk 5]
```diff
@@ -397,19 +397,23 @@ def get_config_directory(section):
     return config if os.path.isdir(config) else os.path.dirname(config)
 
 
-def get_all_bears(log_printer=None, arg_parser=None, silent=True):
+def get_all_bears(log_printer=None,
+                  arg_parser=None,
+                  silent=True,
+                  bear_globs=['**']):
```
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: also here, if possible use a tuple :)

**Comment by shreyans800755** [comment]
> Also some verb in the commit message would be good, like "Add option to ..."

This commit doesn't add the option. Its just using existing option.

> You could in fact leave the tag, as every commit is affecting coala itself ;)

Tag is for the file coala.py and not in the context of whole project.

**Comment by neelanshsahai** [comment]
unack https://github.com/coala/coala/pull/4832/commits/86de9cc9dd3c644d880d40e399cf7049156bd521

**Comment by shreyans800755** [comment]
@neelanshsahai 
Please look at https://github.com/coala/coala/pull/4832#discussion_r149208761.

I've deliberately kept 2 commits as 2nd one of not part of this issue.
And it's already rebased.

**Comment by neelanshsahai** [comment]
:worried: 
Yupp!!
missed that point. Very Well.

**Comment by Makman2** [comment]
@neelanshsahai btw: manual rebase is usually not necessary any more, as we have @gitmate-bot to do that. It's only necessary if there are conflicts which have to be resolved by hand.

**Comment by Makman2** [comment]
Nice I've found a bug in gitmate-bot...

**Comment by shreyans800755** [comment]
Why are manual reviews shown as failed even after pushing new changes or rebasing ?
Ideally it should be based on commit SHA. And its status must be reset after replacing old commit with the new one.

**Comment by Makman2** [comment]
@shreyans800755 gitmate-bot is smart enough to detect code changes. If nothing has changed, it sees it's the same commit even though the SHA is different. Because it was unacked before, it stays unacked. Feel free to discuss that behaviour at [code.gitmate.io](code.gitmate.io), not sure what's better myself :+1:

**Comment by shreyans800755** [comment]
@Makman2 
Not just automated rebased. I've tried to push changes by amending the commit as well before couple of days. But, still it failed both the manual reviews as older equivalent commits were unacked earlier.
I'll discuss this on gitmate.

File: coalib/settings/ConfigurationGathering.py [hunk 6]
```diff
@@ -397,19 +397,23 @@ def get_config_directory(section):
     return config if os.path.isdir(config) else os.path.dirname(config)
 
 
-def get_all_bears(log_printer=None, arg_parser=None, silent=True):
+def get_all_bears(log_printer=None,
+                  arg_parser=None,
+                  silent=True,
+                  bear_globs=tuple('**')):
```
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: this is not what's intended, I believe you thought this creates a single-element tuple with `**` in it, right? ;)
However, `tuple()` collects any iterabel and puts it into a tuple. `string` is an iterable, it allows you to iterate over each char. So in this case you'll get:
`('*', '*')`

It has to be `('**',)` (note the comma)

Comment (Makman2) [reply coalib/settings/ConfigurationGathering.py]: (and without `tuple`, so just `bear_globs=('**',)`)

File: coalib/collecting/Collectors.py [hunk 7]
```diff
@@ -289,12 +289,15 @@ def get_all_bears_names():
...
-def collect_all_bears_from_sections(sections, log_printer=None):
+def collect_all_bears_from_sections(sections,
+                                    log_printer=None,
+                                    bear_globs=tuple('**')):
     """
     Collect all kinds of bears from bear directories given in the sections.
 
     :param sections:    List of sections so bear_dirs are taken into account
     :param log_printer: Log_printer to handle logging
+    :param bear_globs:  List of patterns/bear names to match bears
```
Comment (Makman2) [inline coalib/collecting/Collectors.py]: missing period at end.
\+
What kind of patterns? Mention that these are globs :+1: So just `List of glob patterns ...`

File: coalib/collecting/Collectors.py [hunk 8]
```diff
@@ -289,12 +289,15 @@ def get_all_bears_names():
     return [bear.name for bear in get_all_bears()]
 
 
-def collect_all_bears_from_sections(sections, log_printer=None):
+def collect_all_bears_from_sections(sections,
+                                    log_printer=None,
+                                    bear_globs=tuple('**')):
```
Comment (Makman2) [inline coalib/collecting/Collectors.py]: same here, see above

File: coalib/settings/ConfigurationGathering.py [hunk 9]
```diff
@@ -397,19 +397,23 @@ def get_config_directory(section):
...
     """
     :param log_printer: The log_printer to handle logging.
     :param arg_parser:  An ``ArgParser`` object.
     :param silent:      Whether or not to display warnings.
+    :param bear_globs:  List of patterns/bear names to match bears
```
Comment (Makman2) [inline coalib/settings/ConfigurationGathering.py]: see above

**Comment by Makman2** [comment]
ack 21d285c

**Comment by Makman2** [comment]
unack 4e8221e

**Comment by Makman2** [comment]
Could you maybe write a test that fails when using a `*` as a glob pattern for the default behaviour? So we catch your erroneous default case :+1:

**Comment by shreyans800755** [comment]
> Could you maybe write a test that fails when using a * as a glob pattern for the default behaviour?

Not failing. I replaced default value of `bear_globs` and still tests are passing. Not sure whether to call it bug in glob matching or a feature :smile:

**Comment by shreyans800755** [comment]
One more problem:

Following commands are not giving any bears:
`coala -B --bears *`
`coala -B --bears **`

But,
`coala -B --bears *Clang*`
is giving list of bears having word `Clang` in it.

**Comment by shreyans800755** [comment]
@gitmate-bot rebase

**Comment by shreyans800755** [comment]
@gitmate-bot rebase

**Comment by Makman2** [comment]
> Got the problem. Actually, in first 2 cases, value of args.bears is expanded to list of files in current dir. So, it is not matching in any bears.

Yeah wildcard expansion happens in shell.

> Not failing. I replaced default value of bear_globs and still tests are passing. Not sure whether to call it bug in glob matching or a feature 😄

Wheee that's bad... that's kind of unexpected I believe 😅 However in that case let's file another issue for it to clarify up on that :+1:

File: tests/output/ConsoleInteractionTest.py [hunk 10]
```diff
@@ -1082,3 +1082,25 @@ def test_source_lines(self):
                 {abspath('file'): ('def fun():\n', '    pass  \n')})
             self.assertEqual(stdout.getvalue(),
                              "('    pass  \\n',)\n")
+
+    def test_show_bears_specified_in_args(self):
+        from coalib import coala
+        from tests.TestUtilities import bear_test_module, execute_coala
```
Comment (Makman2) [inline tests/output/ConsoleInteractionTest.py]: why not importing those at top? :O

File: tests/output/ConsoleInteractionTest.py [hunk 11]
```diff
@@ -1082,3 +1082,25 @@ def test_source_lines(self):
...
+    def test_show_bears_specified_in_args_regex(self):
+        from coalib import coala
+        from tests.TestUtilities import bear_test_module, execute_coala
+        with bear_test_module():
+            retval, stdout, stderr = execute_coala(
+                coala.main, 'coala', '-B', '--bears',
+                '*Java*', '*Space*')
+            self.assertEqual(retval, 0)
+            # 2 bear plus 1 line holding the closing colour escape sequence.
+            self.assertEqual(len(stdout.strip().splitlines()), 3)
```
Comment (Makman2) [inline tests/output/ConsoleInteractionTest.py]: I would like to see if we test for the output strings, to see whether the correct bears were matched. If you worry about color, enable `--no-color` in arguments (I hope that's the right flag)

File: tests/output/ConsoleInteractionTest.py [hunk 12]
```diff
@@ -1082,3 +1082,25 @@ def test_source_lines(self):
...
+            # 1 bear plus 1 line holding the closing colour escape sequence.
+            self.assertEqual(len(stdout.strip().splitlines()), 2)
+
+    def test_show_bears_specified_in_args_regex(self):
+        from coalib import coala
+        from tests.TestUtilities import bear_test_module, execute_coala
+        with bear_test_module():
+            retval, stdout, stderr = execute_coala(
+                coala.main, 'coala', '-B', '--bears',
+                '*Java*', '*Space*')
```
Comment (Makman2) [inline tests/output/ConsoleInteractionTest.py]: If possible (though not necessary now):
Can we make this more reliable? As when we add some bear with that name the test will fail.
Maybe it also helps to add a comment explaining what is likely failing so people not involved here immediately understand what has changed.

**Comment by shreyans800755** [comment]
@Makman2 Filed new issue for glob pattern investigation at https://github.com/coala/coala/issues/4878.

File: tests/output/ConsoleInteractionTest.py (hunk 12 above)
Comment (shreyans800755) [reply tests/output/ConsoleInteractionTest.py]: I've modified tests added in this PR not to fail in such cases. I'm testing the name of the bear instead of number of bears now.

File: tests/output/ConsoleInteractionTest.py [hunk 13]
```diff
@@ -1082,3 +1086,20 @@ def test_source_lines(self):
...
+            self.assertEqual(retval, 0)
+            self.assertEqual(stdout.strip(), 'JavaTestBear')
+
+    def test_show_bears_specified_in_args_regex(self):
+        with bear_test_module():
+            retval, stdout, stderr = execute_coala(
+                coala.main, 'coala', '-B', '--bears',
+                '*Java*', '*Space*', '--no-color')
+            self.assertEqual(retval, 0)
+            for bear in stdout.strip().splitlines():
```
Comment (jayvdb) [inline tests/output/ConsoleInteractionTest.py]: This should be changed into an exact match, so the order is locked down and cant randomly change.

Probably want to compare an array of names against
`[l.strip() for l in stdout.strip().splitlines()]`

Comment (shreyans800755) [reply tests/output/ConsoleInteractionTest.py]: I don't think its a good idea.
If someone wants to add a bear matching this criteria, then their changes will break this test if we do an exact match. And it can be difficult to understand the context if the person is newcomer. And this is testing what we are actually doing by running coala.

**Comment by shreyans800755** [comment]
@gitmate-bot rebase

**Comment by shreyans800755** [comment]
@gitmate-bot rebase

File: tests/output/ConsoleInteractionTest.py (hunk 13 above)
Comment (jayvdb) [reply tests/output/ConsoleInteractionTest.py]: Correct.  That is intentional.  Then we can see that new test bears have an impact on the test results.  It should require that the PR adding a test bear adds only one extra line to the output here.  Or they can choose a different name for their test bear.  Not hard.

**Comment by shreyans800755** [comment]
@jayvdb 
I've made requested changes.

> I expect there would be some cli doc changes as part of this PR.

I think its better to put docs changes in coala/documentation. We can have different issue and PR for that.

File: tests/output/ConsoleInteractionTest.py [hunk 14]
```diff
@@ -9,8 +9,10 @@
 
 from testfixtures import LogCapture, StringComparison
 
+
```
Comment (Makman2) [inline tests/output/ConsoleInteractionTest.py]: unnecessary change (and also bad here imo :D)

**Comment by Makman2** [comment]
@shreyans800755 looks good to me (only one style change)
lgtm after that ;)

**Comment by Makman2** [comment]
(Sorry for the late review, was a bit unavailable :3)

**Comment by shreyans800755** [comment]
@Makman2 Made the change and rebased.

**Comment by Makman2** [comment]
ack fa77a8a 8d15735
